Common/Securities/SecurityDefinitionSymbolResolver.cs

```diff
@@ -0,0 +1,163 @@
+using QuantConnect.Interfaces;
+using QuantConnect.Logging;
+using QuantConnect.Util;
+
+namespace QuantConnect.Securities
+{
+    /// <summary>
+    /// Resolves standardized security definitions such as FIGI, CUSIP, ISIN, SEDOL into
+    /// a properly mapped Lean <see cref="Symbol"/>.
+    /// </summary>
+    public class SecurityDefinitionSymbolResolver
+    {
+        private readonly List<SecurityDefinition> _securityDefinitions;
+        private readonly MapFileResolver _mapFileResolver;
+        
+        /// <summary>
+        /// Creates an instance of the symbol resolver
+        /// </summary>
+        /// <param name="dataFolder">Data folder to read the security-database.csv file from</param>
+        /// <param name="mapFileProvider">Map file provider used to obtain symbol mappings</param>
+        /// <param name="market">The market of the assets referred to by the security definitions</param>
+        public SecurityDefinitionSymbolResolver(
+            DirectoryInfo dataFolder = null,
+            IMapFileProvider mapFileProvider = null,
+            string market = Market.USA)
+        {
+            var dataFolderPath = dataFolder?.FullName ?? Globals.DataFolder;
+            
+            var securityDefinitionFile = new FileInfo(
+                Path.Combine(
+                    dataFolderPath,
+                    "symbol-properties",
+                    "security-database.csv"));            
+            
+            if (!SecurityDefinition.TryFromCsvFile(securityDefinitionFile, out _securityDefinitions))
+            {
+                Log.Error($"SecurityDefinitionSymbolResolver(): No security definitions data loaded from file: {securityDefinitionFile.FullName}");
+            }
+
+            if (mapFileProvider == null)
+            {
+                var dataProvider = Composer.Instance.GetExportedValueByTypeName<IDataProvider>(Config.Get("data-provider", "QuantConnect.Lean.Engine.DataFeeds.DefaultDataProvider"));
+                mapFileProvider = Composer.Instance.GetExportedValueByTypeName<IMapFileProvider>(Config.Get("map-file-provider", "LocalDiskMapFileProvider"));
+                mapFileProvider.Initialize(dataProvider);
+            }
+            
+            _mapFileResolver = mapFileProvider.Get(market);
+        }
+        
+        /// <summary>
+        /// Converts CUSIP into a Lean <see cref="Symbol"/>
+        /// </summary>
+        /// <param name="cusip">CUSIP</param>
+        /// <param name="tradingDate">
+        /// The date that the stock was trading at with the CUSIP provided. This is used
+        /// to get the ticker of the symbol on this date.
+        /// </param>
+        /// <returns>The CUSIP's corresponding Symbol</returns>
+        public Symbol CUSIP(string cusip, DateTime tradingDate)
+        {
+            return SecurityDefinitionToSymbol(
+                _securityDefinitions?.FirstOrDefault(x => x.CUSIP == cusip && !string.IsNullOrWhiteSpace(x.CUSIP)),
```

Comment:
`&& !string.IsNullOrWhiteSpace(x.CUSIP)` we should check the argument is not null or whitespace before calling `_securityDefinitions?.FirstOrDefault` and just return null 👍